<a href="https://colab.research.google.com/github/WilliamAshbee/DatascienceHomeworks/blob/main/HW1_Ashbee_William.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Homework 1 - Data Exploration using Census Data

In this homework assignment you will use census data from the 1994 and 1995 current population surveys conducted by the U.S. Census Bureau. A copy of this dataset is located at UCI Machine Learning repository, please see this [link](https://archive.ics.uci.edu/ml/datasets/Census-Income+%28KDD%29) to reach to the dataset website.

This dataset contains census data extracted from the 1994 and 1995 Current Population Surveys. We will only work with 'training' data (a link to that is provided to you below). The data contains 41 demographic and employment related variables. The abbreviated column names provided to you below. You are expected to read the documentation of this dataset, understand the features and preprocess this dataset. Additional information can be found [in the data description](https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.data.html) and [additional comments](https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.names).

Below you will find a code snippet to download and read the data into a pandas dataframe. You can alternatively download it yourself, extract and read it manually. The questions are shown in the subsequent cells. You can provide your answers in this file. Make sure to change the notebook's name and add your name before submitting.

In [1]:
import pandas as pd
import numpy as np

In [2]:
import urllib.request
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.data.gz'
census_dl_file = './census-income.data.gz'
urllib.request.urlretrieve(url, census_dl_file)

('./census-income.data.gz', <http.client.HTTPMessage at 0x7ff2a2258e10>)

In [3]:
columns_abbr = ['AAGE', 'ACLSWKR', 'ADTIND', 'ADTOCC', 'AHGA', 'AHRSPAY', 'AHSCOL', 'AMARITL', 'AMJIND', 'AMJOCC',
                'ARACE', 'AREORGN', 'ASEX', 'AUNMEM', 'AUNTYPE', 'AWKSTAT', 'CAPGAIN', 'CAPLOSS', 'DIVVAL', 
                'FILESTAT', 'GRINREG', 'GRINST', 'HHDFMX', 'HHDREL', 'MARSUPWT', 'MIGMTR1', 'MIGMTR3', 'MIGMTR4', 
                'MIGSAME', 'MIGSUN', 'NOEMP', 'PARENT', 'PEFNTVTY', 'PEMNTVTY', 'PENATVTY', 'PRCITSHP', 'SEOTR', 
                'VETQVA', 'VETYN', 'WKSWORK', 'YEAR', 'PTOTVALB']

In [4]:
df = pd.read_csv('census-income.data.gz', compression='gzip', names=columns_abbr, sep=r',', skipinitialspace=True)
df.head()

AAGE                         ACLSWKR  ADTIND  ...  WKSWORK YEAR  PTOTVALB
0    73                 Not in universe       0  ...        0   95  - 50000.
1    58  Self-employed-not incorporated       4  ...       52   94  - 50000.
2    18                 Not in universe       0  ...        0   95  - 50000.
3     9                 Not in universe       0  ...        0   94  - 50000.
4    10                 Not in universe       0  ...        0   94  - 50000.

[5 rows x 42 columns]

In [5]:
# you can see the DataFrame's info panel here
columns_names = ['AAGE',
                 'ACLSWKR',
                 'ADTIND',
                 'ADTOCC',
                 'AHGA',
                 'AHRSPAY',
                 'AHSCOL', 
                 'AMARITL',
                 'AMJIND', 
                 'AMJOCC',
                'ARACE', 
                 'AREORGN',
                 'ASEX',
                 'AUNMEM', 
                 'AUNTYPE',
                 'AWKSTAT',
                 'CAPGAIN', 
                 'CAPLOSS', 
                 'DIVVAL',
                'FILESTAT',
                 'GRINREG',
                 'GRINST', 
                 'HHDFMX', 
                 'HHDREL', 
                 'MARSUPWT', 
                 'MIGMTR1',
                 'MIGMTR3', 
                 'MIGMTR4',
                'MIGSAME', 
                 'MIGSUN',
                 'NOEMP', 
                 'PARENT', 
                 'PEFNTVTY', 
                 'PEMNTVTY', 
                 'PENATVTY', 
                 'PRCITSHP', 
                 'SEOTR', 
                'VETQVA', 
                 'VETYN', 
                 'WKSWORK', 
                 'YEAR', 
                 'PTOTVALB']

column_desc = {'AAGE':'Age',
'ACLSWKR':'class of worker',
'ADTIND':'industry code',
'ADTOCC':'occupation code',
'AHGA':'education',
'AHRSPAY':'wage per hour',
'AHSCOL':'enrolled in edu inst last wk',
'AMARITL':'marital status',
'AMJIND':'major industry code',
'AMJOCC':'major occupation code',
'ARACE':'mace',
'AREORGN':'hispanic Origin',
'ASEX':'sex',
'AUNMEM':'member of a labor union',
'AUNTYPE':'reason for unemployment',
'AWKSTAT':'full or part time employment stat',
'CAPGAIN':'capital gains',
'CAPLOSS':'capital losses',
'DIVVAL':'divdends from stocks',
'FILESTAT':'tax filer status',
'GRINREG':'region of previous residence',
'GRINST':'state of previous residence',
'HHDFMX':'detailed household and family stat',
'HHDREL':'detailed household summary in household',
'MARSUPWT':'instance weight',
'MIGMTR1':'migration code-change in msa',
'MIGMTR3':'migration code-change in reg',
'MIGMTR4':'migration code-move within reg',
'MIGSAME':'live in this house 1 year ago',
'MIGSUN':'migration prev res in sunbelt',
'NOEMP':'num persons worked for employer',
'PARENT':'family members under 18',
'PEFNTVTY':'country of birth father',
'PEMNTVTY':'country of birth mother',
'PENATVTY':'country of birth self',
'PRCITSHP':'citizenship',
'SEOTR':'own business or self employed',
'VETQVA':'fill inc questionnaire for veterans admin',
'VETYN':'veterans benefits',
'WKSWORK':'weeks worked in year',
'YEAR':'YEAR',
'PTOTVALB':'PTOTVALB'}

a = set(columns_names)
b = set(column_desc.keys())
print(a-b)
print(len(columns_names),len(column_desc))
df.info()

set()
42 42
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199523 entries, 0 to 199522
Data columns (total 42 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   AAGE      199523 non-null  int64  
 1   ACLSWKR   199523 non-null  object 
 2   ADTIND    199523 non-null  int64  
 3   ADTOCC    199523 non-null  int64  
 4   AHGA      199523 non-null  object 
 5   AHRSPAY   199523 non-null  int64  
 6   AHSCOL    199523 non-null  object 
 7   AMARITL   199523 non-null  object 
 8   AMJIND    199523 non-null  object 
 9   AMJOCC    199523 non-null  object 
 10  ARACE     199523 non-null  object 
 11  AREORGN   198649 non-null  object 
 12  ASEX      199523 non-null  object 
 13  AUNMEM    199523 non-null  object 
 14  AUNTYPE   199523 non-null  object 
 15  AWKSTAT   199523 non-null  object 
 16  CAPGAIN   199523 non-null  int64  
 17  CAPLOSS   199523 non-null  int64  
 18  DIVVAL    199523 non-null  int64  
 19  FILESTAT  199523 non-null  objec

### Q1 - Identify the data scales and data types for each variable in census data. Identify the domain for each variable by checking the attributes' values. Then, create a data quality report for both categorical (nominal, ordinal) and continuous (interval, ratio) variables. [40 pts]

For data scales, identify whether an attribute is nominal, ordinal, interval or ratio scale.
For data types, identify the domain and provide an appropriate data type (integer, float, String, date, Boolean). See if that data type is correct in your dataframe.
For domain, inspect each distinct value for each attribute. Identify missing values. Also include the bar plots and histograms for visualizing the distributions. 

*The examples for a continious and a categorical feature can be seen below. You do not need to use jupyter formatting provided here. You can print a DataFrame or read a csv, and display it. Make sure you have the csv in your zip file.*

In [6]:
# your code goes here




In [7]:
#df.sum()
# We will use df.iteritems() as it allows us to iterate through columns
for (name, series) in df.iteritems():
  print('ANALYZING THE COLUMN:', name)
  print('\tTotal number of records', series.size)
  print('\tNumber of missing values', series.isnull().sum())
  print('\tPercentage of missing values {0}%'.format(((series.isnull().sum()/series.size)*100)) )
  print('\tNumber of unique values', series.unique().size)
  print('| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |')
  
  catagorical = True
  if catagorical:

    modes= series.value_counts().keys()
    card = series.size
    mode1= modes[0]
    mode2= modes[1]
    mode1count= series.value_counts()[mode1]
    mode2count= series.value_counts()[mode2]
    mode1percent = "{:.2f}".format(mode1count*100.0/float(card))
    mode2percent = "{:.2f}".format(mode2count*100.0/float(card))
  
    print(name,"|",
          column_desc[name],"|",
          series.size,"|",
          series.isnull().sum(),"|",
          series.unique().size,"|",
          mode1,"|",mode1count,"|",mode1percent,"|",
          mode2,"|",mode2count,"|",mode2percent
          )
    
  #print(mode1,mode1count,mode1percent,mode2,mode2count,mode2percent)


ANALYZING THE COLUMN: AAGE
	Total number of records 199523
	Number of missing values 0
	Percentage of missing values 0.0%
	Number of unique values 91
| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |
AAGE | Age | 199523 | 0 | 91 | 34 | 3489 | 1.75 | 35 | 3450 | 1.73
ANALYZING THE COLUMN: ACLSWKR
	Total number of records 199523
	Number of missing values 0
	Percentage of missing values 0.0%
	Number of unique values 9
| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |
ACLSWKR | class of worker | 199523 | 0 | 9 | Not in universe | 100245 | 50.24 | Private | 72028 | 36.10
ANALYZING THE COLUMN: ADTIND
	Total number of records 199523
	Number of missing values 0
	Percentage of missing values 0.0%
	Number of unique values 52
| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd

In [20]:
for (name, series) in df.iteritems():
  
  catagorical = True
  if catagorical:

    
    try:
      min = '{}'.format(series.min())
      q1 = '{}'.format(series.quantile(0.25))
      median = '{}'.format(series.median())
      q3 = '{}'.format(series.quantile(0.75))
      max = '{}'.format(series.max())
      mean = '{:.2f}'.format(series.mean())
      stdev = '{:.2f}'.format(series.std()) 

      #| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
      
      print(name,"|",
            column_desc[name],"|",
            series.size,"|",
            series.isnull().sum(),"|",
            series.unique().size,"|",
            min,"|",
            q1,"|",
            median, "|",
            q3,"|",
            max,"|",
            mean,"|",
            stdev,"|"
            )
    except:
        print(series.name, ' is not continuous')

AAGE | Age | 199523 | 0 | 91 | 0 | 15.0 | 33.0 | 50.0 | 90 | 34.49 | 22.31 |
ACLSWKR  is not continuous
ADTIND | industry code | 199523 | 0 | 52 | 0 | 0.0 | 0.0 | 33.0 | 51 | 15.35 | 18.07 |
ADTOCC | occupation code | 199523 | 0 | 47 | 0 | 0.0 | 0.0 | 26.0 | 46 | 11.31 | 14.45 |
AHGA  is not continuous
AHRSPAY | wage per hour | 199523 | 0 | 1240 | 0 | 0.0 | 0.0 | 0.0 | 9999 | 55.43 | 274.90 |
AHSCOL  is not continuous
AMARITL  is not continuous
AMJIND  is not continuous
AMJOCC  is not continuous
ARACE  is not continuous
AREORGN  is not continuous
ASEX  is not continuous
AUNMEM  is not continuous
AUNTYPE  is not continuous
AWKSTAT  is not continuous
CAPGAIN | capital gains | 199523 | 0 | 132 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 434.72 | 4697.53 |
CAPLOSS | capital losses | 199523 | 0 | 113 | 0 | 0.0 | 0.0 | 0.0 | 4608 | 37.31 | 271.90 |
DIVVAL | divdends from stocks | 199523 | 0 | 1478 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 197.53 | 1984.16 |
FILESTAT  is not continuous
GRINREG  is not continuous
G

| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| AAGE | Age | 199,523 | 0 | 91 | 0 | 15 | 33 | 50 | 90 | 34.49 | 22.31 |  |

1.   List item

*   List item
*   List item

---




2.   List item

| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| ACLSWKR | Class of worker | 199,523 | 0 | 9 | Not in Universe | 100,245 | 50.24 | Private | 72,028 |  36.10 |  | *italicized text*

In [8]:
#(df['AAGE']==32).sum()#32,73
modes= df['AAGE'].value_counts().keys()
card = df['AAGE'].size
mode1= modes[0]
mode2= modes[1]
mode1count= df['AAGE'].value_counts()[mode1]
mode2count= df['AAGE'].value_counts()[mode2]
mode1percent = "{:.2f}".format(mode1count*100.0/float(card))
mode2percent = "{:.2f}".format(mode2count*100.0/float(card))
print(mode1,mode1count,mode1percent,mode2,mode2count,mode2percent)

34 3489 1.75 35 3450 1.73


In [16]:
series = df['AAGE']
card = df['AAGE'].size
min = '{}'.format(series.min())
q1 = '{}'.format(series.quantile(0.25))
median = '{}'.format(series.median())
q3 = '{}'.format(series.quantile(0.75))
max = '{}'.format(series.max())
mean = '{:.2f}'.format(series.mean())
stdev = '{:.2f}'.format(series.std()) 
#var = '{:.4f}'.format(series.var())

#| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
print(card,"|",
      min,"|",
      q1,"|",
      median, "|",
      q3,"|",
      max,"|",
      mean,"|",
      stdev,"|"
      )

199523 | 0 | 15.0 | 33.0 | 50.0 | 90 | 34.49 | 22.31 |


### Data Quality Report for Cont Variables

| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| AAGE | Age | 199,523 | 0 | 91 | 0 | 15 | 33 | 50 | 90 | 34.49 | 22.31 |  |

1.   List item

*   List item
*   List item


2.   List item



### Data Quality Report for Categorical Variables

| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| ACLSWKR | Class of worker | 199,523 | 0 | 9 | Not in Universe | 100,245 | 50.24 | Private | 72,028 |  36.10 |  | *italicized text*

### Q2 - Identify the data quality issues for each attribute. Create a data quality plan and implement the changes. [35 pts] 

Identify missing values, potential anomalies and outliers. Finally, remedy these issues.

In [9]:
# your code goes here

*italicized text*### Q3 - For all the interval and ratio scale features, create a scatter plot matrix (pairplots) and correlation matrix. For all ordinal scale features, create a correlation matrix using Spearman correlation. [35 pts]

Use the data scales list you created in Q1.

In [10]:
# your code goes here